src/lib/udunits.3
." $Id$
." $__Header$
.TH UDUNITS 3 "$Date$" "Printed: \n(yr.\n(mo.\n(dy" "UNIDATA LIBRARY FUNCTIONS"
.SH NAME
udunits, utInit, utScan, utCalendar, utInvCalendar, utConvert, utPrint, utIsTime, utHasOrigin, utClear, utCopy, utScale, utDivide, utInvert, utMultiply, utRaise, utTerm \- Unidata units library
.SH SYNOPSIS
.nf
.ft B
cc -I\fIunidata_inc\fP ...
.sp
#include <udunits.h>
.sp
.ta 4m +\w'const utUnit 'u
int utInit(
	const char	*path);
.sp
int utScan(
	const char	*spec,
	utUnit	*unit);
.sp
int utCalendar(
	double	value;
	utUnit	*unit,
	int	*year,
	int	*month,
	int	*day,
	int	*hour,
	int	*minute,
	float	*second);
.sp
int utInvCalendar(
	int	year,
	int	month,
	int	day,
	int	hour,
	int	minute,
	double	second,
	utUnit	*unit,
	double	*value);
.sp
int utConvert(
	const utUnit	*from,
	const utUnit	*to,
	double	*slope,
	double	*intercept);
.sp
int *utPrint(
	const utUnit	*unit,
	char		**s);
.sp
utUnit *utClear(
	utUnit	*unit);
.sp
int *utIsTime(
	const utUnit	*unit);
.sp
int *utHasOrigin(
	const utUnit	*unit);
.sp
utUnit *utCopy(
	const utUnit	*source,
	utUnit	*dest);
.sp
utUnit *utScale(
	utUnit	*source,
	double	factor,
	utUnit	*result);
.sp
utUnit *utDivide(
	utUnit	*numer,
	utUnit	*denom, 
	utUnit	*result);
.sp
utUnit *utInvert(
	utUnit	*source,
	utUnit	*result);
.sp
utUnit *utMultiply(
	utUnit	*term1,
	utUnit	*term2, 
	utUnit	*result);
.sp
utUnit *utRaise(
	utUnit	*source,
	int	power,
	utUnit	*result);
.sp
void utTerm();
.ft
.fi
.SH DESCRIPTION
.LP
The Unidata units library, \fBudunits\fP, supports conversion of unit
specifications between formatted and binary forms, arithmetic
manipulation of unit specifications, and conversion of values between 
compatible scales of measurement.
.LP
A unit is the amount by which a physical quantity is measured.  For example:
.sp
.TS
center tab(|);
c c
c c.
Physical Quantity|Possible Unit
\_|\_
time|weeks
distance|centimeters
power|watts
.TE
.sp
A unit can have an origin associated with it \-\- in which case
the unit and origin together define a scale.  For example, the
phrase "the temperature is 25 degrees Celsius" specifies a particular point on a
measurement scale; whereas the phrase "the temperature difference is
25 degrees Celsius" specifies a unit with no origin and, hence, no associated
scale.  If not remembered, this subtle distinction can cause problems when
handling units.
.LP
.B utInit()
initializes the units package.  If \fIpath\fP is non-NULL and not empty,
then it specifies a units file containing initializing unit definitions;
otherwise, the environment variable \fBUDUNITS_PATH\fR is checked and,
if it exists and is not empty, then it is assumed to contain the pathname
of the units file;
otherwise, a compile-time default pathname is used.
.sp
The definitions in the units file are read into memory.
This function returns 
0 on success, 
UT_ENOFILE if the units file doesn't exist,
UT_ESYNTAX if the units file contains a syntax error,
UT_EUNKNOWN if the units file contains an unknown specification,
UT_EIO if an I/O error occurred while accessing the units file, and
UT_EALLOC if a memory allocation failure occurred.
.LP
.B utScan()
converts the formatted unit specification \fIspec\fP into a binary
unit structure and stores the result in \fIunit\fP.
The binary representation is used for algebraic manipulation.
This function returns 0 on success,
UT_ENOINIT if the package hasn't been initialized,
UT_EINVALID if the unit argument is a null pointer,
UT_EUNKNOWN if the specification contains an unknown unit, and
UT_ESYNTAX if the specification contains a syntax error.
.LP
.B utCalendar()
converts the amount, \fIvalue\fP, of the temporal unit, \fIunit\fP,
into a UTC-referenced date and time 
(see, however, the section on HANDLING TIME).
The reference unit shall be a time unit and have an origin.
This function returns 0 on success,
UT_ENOINIT if the package hasn't been initialized and
UT_EINVALID if the unit structure is not a temporal one.
.LP
.B utInvCalendar()
converts a UTC-referenced date and time
into the amount, \fIvalue\fP, of the temporal unit, \fIunit\fP
(see, however, the section on HANDLING TIME).
The reference unit shall be a time unit and have an origin.
This function returns 0 on success,
UT_ENOINIT if the package hasn't been initialized and
UT_EINVALID if the unit structure is not a temporal one.
.LP
.B utConvert()
returns the coefficients of the Galilean transformation (i.e. y = a*x + b)
necessary to convert the \fIfrom\fP unit into the \fIto\fP unit.
The units must be compatible (i.e., their quotient must be dimensionless).
On successful return, \fIslope\fP and \fIintercept\fP will contain the 
values for the slope and intercept coefficients, respectively.
This function returns
0 on success,
UT_ENOINIT if the package hasn't been initialized,
UT_EINVALID if one of the unit structures is invalid, and
UT_ECONVERT if the units are not convertible.
.LP
.B utPrint()
converts the binary unit structure \fIunit\fP
into a formatted unit specification and stores the string into a static
buffer.  
The argument \fIs\fP is set to point to the static buffer.
The string should not be modified (it may, however, be copied)
and will be overwritten by subsequent calls to this function.
This function returns
0 on success,
UT_ENOINIT if the package hasn't been initialized, and
UT_EINVALID if the unit structure is invalid.
On error, the string argument is set to NULL.
.LP
.B utClear()
clears a unit structure by setting it to the dimensionless value 1.
.LP
.B utIsTime()
returns true if the given unit structure refers to a time unit and false
otherwise.
This function ignores whether or not the unit has an origin.
.LP
.B utHasOrigin()
returns true of the given unit structure has an origin (i.e. defines a
scale) and false otherwise.
.LP
.B utCopy()
copies the unit structure \fIsource\fP to the unit structure
\fIdest\fP.
This function correctly handles the case where the same unit structure
is referenced by the source and destination units.
The address of the destination unit structure is returned.
.LP
.B utScale()
scales the unit structure \fIsource\fP by 
\fIfactor\fP, storing the result in the unit structure \fIresult\fP.
This function correctly handles the case where the same unit structure
is referenced by the source and result units.
The address of the result unit structure is returned.
.LP
.B utInvert()
inverts the unit structure \fIsource\fP, storing the result in
unit structure \fIresult\fP.
Multiplying a unit by its reciprocal yields the
dimensionless value 1.
This function correctly handles the case where the source and result
unit refer to the same structure.
If successful, this function returns the address of the result
unit structure; otherwise, it returns NULL.
.LP
.B utDivide()
divides unit structure \fInumer\fP by unit structure \fIdenom\fP
and stores the result in unit structure \fIresult\fP.
This function correctly handles the case where the same unit structure
is referenced by two or more arguments.
If successful, this function returns the address of the result
unit structure; otherwise, it returns NULL.
.LP
.B utMultiply()
multiplies unit structure \fIterm1\fP by unit structure \fIterm2\fP
and stores the result in unit structure \fIresult\fP.
This function correctly handles the case where the same unit structure
is referenced by two or more arguments.
If successful, this function returns the address of the result
unit structure; otherwise, it returns NULL.
.LP
.B utRaise()
raises the unit structure \fIsource\fP by the power \fPpower\fP,
storing the result in the unit structure \fIresult\fP.
This function correctly handles the case where the same unit structure
is referenced by the source and result units.
If successful, this function returns the address of the result
unit structure; otherwise, it returns NULL.
.LP
.B utTerm()
terminates usage of this package.  In particular, it frees all allocated
memory.  It should be called when the library is no longer needed.
.SH "HANDLING TIME"
.LP
The \fBudunits\fP(3) package uses a mixed Gregorian/Julian calendar
system.
Dates prior to 1582-10-15 are assumed to use the Julian calendar, which
was introduced by Julius Caesar in 46 BCE and is based on a year that 
is exactly 365.25 days long.
Dates on and after 1582-10-15 are assumed to use the Gregorian calendar,
which was introduced on that date and is based on a year that is
exactly 365.2425 days long.
(A year is actually approximately 365.242198781 days long.)
Seemingly strange behavior of the \fBudunits\fP(3) package can
result if a user-given time interval includes the changeover date.
For example, \fButCalendar\fP() and \fButInvCalendar\fP() can be used
to show that 1582-10-15 *preceeded* 1582-10-14 by 9 days.
.SH "EXAMPLES"
.LP
Convert two data sets to
a common unit, subtract one from the other, then
save the result in a (different) output unit:
.sp
.RS +4
.nf
.ta 4m +\w'utUnit 'u +\w'double 'u +4m +4m
if (utInit("") != 0) {
	/* handle initialization error */
} else {
	char	*UnitString1, *UnitString2, *OutputUnitString;
	utUnit	unit1, unit2, OutputUnit;
	...
	if (utScan(UnitString1, &unit1) != 0 || 
		utScan(UnitString2, &unit2) != 0 ||
		utScan(OutputUnitString2, &OutputUnit) != 0) {
		/*
		 * handle decode error
		 */
	} else {
		double	InSlope, InIntercept;
		double	OutSlope, OutIntercept;
		...
		if (utConvert(&unit2, &unit1, &InSlope, 
			&InIntercept) != 0 ||
		    utConvert(&unit1, &OutputUnit, &OutSlope, 
			&OutIntercept) != 0) {
			/*
			 * handle data-incompatibility
			 */
		} else {
			/*
			 * process data using:
			 *	OutputValue = OutSlope*(Data1Value -
			 *		(InSlope*Data2Value + InIntercept)) 
			 *		+ OutIntercept
			 */
		}
	}
	utTerm();
}
.fi
.RE
.LP
the above example could be made more efficient
by testing the returned conversion factors for nearness to 1 and 0 and using
appropriately streamlined processing expressions.
.sp
.LP
Compute a threshold value corresponding to an input data value plus a
user-specified delta (the units of the input data value and delta can
differ):
.sp
.RS +4
.nf
.ta 4m +\w'utUnit 'u +\w'double 'u +4m +4m
char	*input_unit_string, *delta_unit_string;
float	input_value;
utUnit	input_unit, delta_unit;
\&...
if (utScan(input_unit_string, &input_unit) != 0 || 
	utScan(delta_unit_string, &delta_unit) != 0) {
	/*
	 * handle decode error
	 */
} else {
	double	slope, intercept;

src/RELEASE_NOTES
VERSION		CHANGE

1.11.7	1998-11-25
		Ported to the following systems:
		    AIX 4.3 (both /bin/c89 and /usr/vac/bin/c89)
		    HP-UX B.11.00.

1.11.6	1998-02-10
		Corrected definitions of "perm_0C" and "perm_23C" in the
		database.  The definitions used "Pa" before it was
		defined as "pascal".

		Added "L" as an abbreviation for "liter".

		Added declaration of the Fortran function UTICALTIME to
		the header-file "udunits.inc".

		Changed installation of manual pages on HP-UX systems: 
		the manual page is no longer preprocessed with tbl(1)
		and neqn(1) because the man(1) utility incorporates
		that.

                Corrected bug in Fortran UTENC() function: a blank
                output character variable no longer results in a
                UT_ENOROOM return status.

1.11.5	1997-08-18
		Corrected return value of valtocal(): it now matches the 
		specification (returns 0 on success).  Corrected "test.pl"
		to match.

		Corrected "test.pl" to not test clear() return value since 
		it doesn't have one.

1.11.4	1997-06-10
		Removed `-traditional' from gcc(1) example in INSTALL
		and CUSTOMIZE files because it doesn't work.

		Fixed bug in the dectime() of lib/utlib.c that caused 
		timestamps around 1925-01-01 (and erlier) to be 
		misinterpreted.

                Added caution to file INSTALL about building the perl
                extension module using the same environment that built
                the perl package.

		Modified FORTRAN header file "udunits.inc": should
		be more robust now and work under more compilers.

                Made more explicit in the README file what needs to be
                included in a request for help.

		Fixed udunits(3f) manual page: included missing file
		"testcal2.f" in the distribution and modified the
		build procedure to make "lib/udunits.3f" from
		"lib/udunits.3fi" and "lib/testcal2.f".

		Added missing "meters" to definition of "light_year"

		Ported to UNICOS-9/J90 (including Perl extension module).
		Something seems to be wrong with installing the new, static
		perl(1) utility, however.

		Ported to CFORTRAN version 3.9.

1.11.3	1996-12-2
                Corrected and updated the units database.  There were
                three major corrections due to typo's:
		
		    Unit		Was		Now Is
		    ----		---		------
		    assay_ton		2.916667e2 kg	2.916667e-2 kg
		    register_ton	3.831685 m3	2.831685 m3
		    pond		1.806650e-3 N	9.806650e-3 N
		
                Other modifications were basically updates to the
                conversion factors based on the latest information from
                the National Institute for Standards and Technology
                (NIST).  The database was originally bsed on the 1991
                "Guide for the Use of the International System of Units
                (SI)"; the more recent information is based on the 1995
                publication by the same name.  See the following URL for
                more information than you could possibly want:

		    http://physics.nist.gov/Divisions/Div840/SI.html

1.11.2	1996-11-15
		Improved accounting for finite-precision in computations
		involving time.  Should eliminate incorrect utCalendar()
		returns when the time is an exact multiple of an
		integral number of seconds.

1.11.1	1996-11-14
                Improved support for generating `whatis' database:
                switched from use of catman(1)'s `-M' option to use of
                MANPATH environment variable.  This adds creation of the
                manual-page database to AIX and should provide better
                support for Linix.

                Added missing files (port/misc/udalloc.{c,h}) to
                distribution (sorry 'bout that) and tested installation
                from scratch.

                Improved accounting for finite-precision in computations
                involving time.  Should eliminate utCalendar() returns
                with 59 minutes and 60 seconds.

		Added function prototype for utInvCalendar() to UDUNITS
		header file (udunits.h).

1.11	1996-11-04
                Converted support for FORTRAN-calling-C from the fortc(1)
		utility to the `cfortran.h' header-file.

		Modified installation procedure to install the FORTRAN
		include file mentioned in the manual page (udunits.inc).

		Corrected definition of the International Table British
		Thermal Unit (It_Btu) from 1.055056 joules to 1.055056e3
		joules).

		Corrected error message when the database contains too
		many base units.

		Added `count' as a base unit.  Moved `bit' to a non-base
		unit of `count'.

		Added routine declarations to FORTRAN header-file
		(udunits.inc).

		Corrected spelling of `termporal' in the perl
		test-script.

		Fixed memory-leak in utTerm() (UTCLS) routine.

                Added ignoring of trailing whitespace in unit
                specifications.

1.10.8	1996-06-27
		Added missing file (port/fortc/bsdos.m4) to distribution.

1.10.7	1996-05-02
                Ported to BSD/OS 2.1 (but with no FORTRAN or perl
		support).

1.10.6	1996-04-23
		Added missing tfind() to package-supplied, portability
		functions (port/misc/tfind.c).

1.10.5	1996-04-19

src/perl/UDUNITS.c
/*
 * This file was generated automatically by xsubpp version 1.9507 from the 
 * contents of UDUNITS.xs. Do not edit this file, edit UDUNITS.xs instead.
 *
 *	ANY CHANGES MADE HERE WILL BE LOST! 
 *
 */

#line 1 "UDUNITS.xs"
#include <stdio.h>	/* for diagnostics */
#include <stdlib.h>
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#include "udunits.h"

extern utUnit	*utNew();

static int
not_here(s)
char *s;
{
    croak("%s not implemented on this architecture", s);
    return -1;
}

static double
constant(name, arg)
char *name;
int arg;
{
    errno = 0;
    switch (*name) {
    case 'A':
	break;
    case 'B':
	break;
    case 'C':
	break;
    case 'D':
	break;
    case 'E':
	break;
    case 'F':
	break;
    case 'G':
	break;
    case 'H':
	break;
    case 'I':
	break;
    case 'J':
	break;
    case 'K':
	break;
    case 'L':
	break;
    case 'M':
	break;
    case 'N':
	break;
    case 'O':
	break;
    case 'P':
	break;
    case 'Q':
	break;
    case 'R':
	break;
    case 'S':
	break;
    case 'T':
	break;
    case 'U':
	if (strEQ(name, "EALLOC"))
#ifdef UT_EALLOC
	    return UT_EALLOC;
#else
	    goto not_there;
#endif
	if (strEQ(name, "ECONVERT"))
#ifdef UT_ECONVERT
	    return UT_ECONVERT;
#else
	    goto not_there;
#endif
	if (strEQ(name, "EINVALID"))
#ifdef UT_EINVALID
	    return UT_EINVALID;
#else
	    goto not_there;
#endif
	if (strEQ(name, "EIO"))
#ifdef UT_EIO
	    return UT_EIO;
#else
	    goto not_there;
#endif
	if (strEQ(name, "ENOFILE"))
#ifdef UT_ENOFILE
	    return UT_ENOFILE;
#else
	    goto not_there;
#endif
	if (strEQ(name, "ENOINIT"))
#ifdef UT_ENOINIT
	    return UT_ENOINIT;
#else
	    goto not_there;
#endif
	if (strEQ(name, "ENOROOM"))
#ifdef UT_ENOROOM
	    return UT_ENOROOM;
#else
	    goto not_there;
#endif
	if (strEQ(name, "ENOTTIME"))
#ifdef UT_ENOTTIME
	    return UT_ENOTTIME;
#else
	    goto not_there;
#endif
	if (strEQ(name, "EOF"))
#ifdef UT_EOF
	    return UT_EOF;
#else
	    goto not_there;
#endif
	if (strEQ(name, "ESYNTAX"))
#ifdef UT_ESYNTAX
	    return UT_ESYNTAX;
#else
	    goto not_there;
#endif
	if (strEQ(name, "EUNKNOWN"))
#ifdef UT_EUNKNOWN
	    return UT_EUNKNOWN;
#else
	    goto not_there;
#endif
	if (strEQ(name, "MAXNUM_BASE_QUANTITIES"))
#ifdef UT_MAXNUM_BASE_QUANTITIES
	    return UT_MAXNUM_BASE_QUANTITIES;
#else
	    goto not_there;
#endif
	if (strEQ(name, "NAMELEN"))
#ifdef UT_NAMELEN
	    return UT_NAMELEN;
#else
	    goto not_there;
#endif
	break;
    case 'V':
	break;
    case 'W':
	break;
    case 'X':
	break;
    case 'Y':
	break;
    case 'Z':
	break;
    }
    errno = EINVAL;
    return 0;

not_there:
    errno = ENOENT;
    return 0;
}


#line 176 "UDUNITS.c"
XS(XS_UDUNITS_constant)
{
    dXSARGS;
    if (items != 2)
	croak("Usage: UDUNITS::constant(name,arg)");
    {
	char *	name = (char *)SvPV(ST(0),PL_na);
	int	arg = (int)SvIV(ST(1));
	double	RETVAL;

	RETVAL = constant(name, arg);
	ST(0) = sv_newmortal();
	sv_setnv(ST(0), (double)RETVAL);
    }
    XSRETURN(1);
}

XS(XS_UDUNITS_init)
{
    dXSARGS;
    if (items != 1)
	croak("Usage: UDUNITS::init(path)");
    {
	char *	path = (char *)SvPV(ST(0),PL_na);
	int	RETVAL;
#line 179 "UDUNITS.xs"
    {
	RETVAL = utInit(path);
    }
#line 206 "UDUNITS.c"
	ST(0) = sv_newmortal();
	sv_setiv(ST(0), (IV)RETVAL);
    }
    XSRETURN(1);
}

XS(XS_UDUNITS_term)
{
    dXSARGS;
    if (items != 0)
	croak("Usage: UDUNITS::term()");
    {
#line 189 "UDUNITS.xs"
    {
	utTerm();
    }
#line 223 "UDUNITS.c"
    }
    XSRETURN_EMPTY;
}

XS(XS_UDUNITS_new)
{
    dXSARGS;
    if (items != 0)
	croak("Usage: UDUNITS::new()");
    {
	utUnit *	RETVAL;
#line 197 "UDUNITS.xs"
    {
	RETVAL = utNew();

	if (RETVAL == 0)
	    croak("Couldn't allocate %lu bytes for new unit structure",
		  sizeof(utUnit));
    }
#line 243 "UDUNITS.c"
	ST(0) = sv_newmortal();
	sv_setref_pv(ST(0), "utUnitPtr", (void*)RETVAL);
    }
    XSRETURN(1);
}

XS(XS_UDUNITS_scan)
{
    dXSARGS;
    if (items != 1)
	croak("Usage: UDUNITS::scan(spec)");
    {
	char *	spec = (char *)SvPV(ST(0),PL_na);
#line 212 "UDUNITS.xs"
    {
	int	status;
	utUnit	*unit = utNew();

	if (unit == 0)
	    croak("Couldn't allocate %lu bytes for new unit structure",
		  sizeof(utUnit));

	ST(0) = sv_newmortal();

	status = utScan(spec, unit);

	if (status == UT_ENOINIT)
	    croak("units module not initialized");
	else if (status == 0)
	    sv_setref_pv(ST(0), "utUnitPtr", (void*)unit);
    }
#line 275 "UDUNITS.c"
    }
    XSRETURN(1);
}

XS(XS_utUnitPtr_istime)
{
    dXSARGS;
    if (items != 1)
	croak("Usage: utUnitPtr::istime(unit)");
    {
	utUnit *	unit;
	int	RETVAL;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 238 "UDUNITS.xs"
    {
	RETVAL = utIsTime(unit);
    }
#line 299 "UDUNITS.c"
	ST(0) = sv_newmortal();
	sv_setiv(ST(0), (IV)RETVAL);
    }
    XSRETURN(1);
}

XS(XS_utUnitPtr_hasorigin)
{
    dXSARGS;
    if (items != 1)
	croak("Usage: utUnitPtr::hasorigin(unit)");
    {
	utUnit *	unit;
	int	RETVAL;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 249 "UDUNITS.xs"
    {
	RETVAL = utHasOrigin(unit);
    }
#line 325 "UDUNITS.c"
	ST(0) = sv_newmortal();
	sv_setiv(ST(0), (IV)RETVAL);
    }
    XSRETURN(1);
}

XS(XS_utUnitPtr_clear)
{
    dXSARGS;
    if (items != 1)
	croak("Usage: utUnitPtr::clear(unit)");
    {
	utUnit *	unit;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 260 "UDUNITS.xs"
    {
	(void) utClear(unit);
    }
#line 350 "UDUNITS.c"
    }
    XSRETURN_EMPTY;
}

XS(XS_utUnitPtr_dup)
{
    dXSARGS;
    if (items != 1)
	croak("Usage: utUnitPtr::dup(source)");
    {
	utUnit *	source;
	utUnit *	RETVAL;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    source = (utUnit *) tmp;
	}
	else
	    croak("source is not of type utUnitPtr");
#line 269 "UDUNITS.xs"
    {
	utUnit	*dest = utNew();

	if (dest == 0)
	    croak("Couldn't allocate %lu bytes for new unit structure",
		  sizeof(utUnit));

	RETVAL = utCopy(source, dest);
    }
#line 380 "UDUNITS.c"
	ST(0) = sv_newmortal();
	sv_setref_pv(ST(0), "utUnitPtr", (void*)RETVAL);
    }
    XSRETURN(1);
}

XS(XS_utUnitPtr_shift)
{
    dXSARGS;
    if (items != 2)
	croak("Usage: utUnitPtr::shift(unit, amount)");
    {
	utUnit *	unit;
	double	amount = (double)SvNV(ST(1));

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 287 "UDUNITS.xs"
    {
	(void) utShift(unit, amount, unit);
    }
#line 406 "UDUNITS.c"
    }
    XSRETURN_EMPTY;
}

XS(XS_utUnitPtr_scale)
{
    dXSARGS;
    if (items != 2)
	croak("Usage: utUnitPtr::scale(unit, coefficient)");
    {
	utUnit *	unit;
	double	coefficient = (double)SvNV(ST(1));

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 297 "UDUNITS.xs"
    {
	(void) utScale(unit, coefficient, unit);
    }
#line 430 "UDUNITS.c"
    }
    XSRETURN_EMPTY;
}

XS(XS_utUnitPtr_multiply)
{
    dXSARGS;
    if (items != 2)
	croak("Usage: utUnitPtr::multiply(unit, otherunit)");
    {
	utUnit *	unit;
	utUnit *	otherunit;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");

	if (sv_derived_from(ST(1), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(1)));
	    otherunit = (utUnit *) tmp;
	}
	else
	    croak("otherunit is not of type utUnitPtr");
#line 307 "UDUNITS.xs"
    {
	(void) utMultiply(unit, otherunit, unit);
    }
#line 461 "UDUNITS.c"
    }
    XSRETURN_EMPTY;
}

XS(XS_utUnitPtr_invert)
{
    dXSARGS;
    if (items != 1)
	croak("Usage: utUnitPtr::invert(unit)");
    {
	utUnit *	unit;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 316 "UDUNITS.xs"
    {
	(void) utInvert(unit, unit);
    }
#line 484 "UDUNITS.c"
    }
    XSRETURN_EMPTY;
}

XS(XS_utUnitPtr_divide)
{
    dXSARGS;
    if (items != 2)
	croak("Usage: utUnitPtr::divide(unit, divisor)");
    {
	utUnit *	unit;
	utUnit *	divisor;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");

	if (sv_derived_from(ST(1), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(1)));
	    divisor = (utUnit *) tmp;
	}
	else
	    croak("divisor is not of type utUnitPtr");
#line 326 "UDUNITS.xs"
    {
	(void) utDivide(unit, divisor, unit);
    }
#line 515 "UDUNITS.c"
    }
    XSRETURN_EMPTY;
}

XS(XS_utUnitPtr_raise)
{
    dXSARGS;
    if (items != 2)
	croak("Usage: utUnitPtr::raise(unit, power)");
    {
	utUnit *	unit;
	int	power = (int)SvIV(ST(1));

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 336 "UDUNITS.xs"
    {
	(void) utRaise(unit, power, unit);
    }
#line 539 "UDUNITS.c"
    }
    XSRETURN_EMPTY;
}

XS(XS_utUnitPtr_print)
{
    dXSARGS;
    if (items != 1)
	croak("Usage: utUnitPtr::print(unit)");
    {
	utUnit *	unit;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 345 "UDUNITS.xs"
    {
	char	*staticbuf;

	(void) utPrint(unit, &staticbuf);
	ST(0) = sv_newmortal();
	sv_setpv((SV*)ST(0), staticbuf);
    }
#line 566 "UDUNITS.c"
    }
    XSRETURN(1);
}

XS(XS_utUnitPtr_convert)
{
    dXSARGS;
    if (items != 4)
	croak("Usage: utUnitPtr::convert(from_unit, to_unit, slope, intercept)");
    {
	utUnit *	from_unit;
	utUnit *	to_unit;
	double	slope;
	double	intercept;
	int	RETVAL;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    from_unit = (utUnit *) tmp;
	}
	else
	    croak("from_unit is not of type utUnitPtr");

	if (sv_derived_from(ST(1), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(1)));
	    to_unit = (utUnit *) tmp;
	}
	else
	    croak("to_unit is not of type utUnitPtr");
#line 361 "UDUNITS.xs"
    {
	RETVAL = utConvert(from_unit, to_unit, &slope, &intercept);
    }
#line 600 "UDUNITS.c"
	sv_setnv(ST(2), (double)slope);
	SvSETMAGIC(ST(2));
	sv_setnv(ST(3), (double)intercept);
	SvSETMAGIC(ST(3));
	ST(0) = sv_newmortal();
	sv_setiv(ST(0), (IV)RETVAL);
    }
    XSRETURN(1);
}

XS(XS_utUnitPtr_valtocal)
{
    dXSARGS;
    if (items != 8)
	croak("Usage: utUnitPtr::valtocal(unit, value, year, month, day, hour, minute, second)");
    {
	utUnit *	unit;
	double	value = (double)SvNV(ST(1));
	int	year;
	int	month;
	int	day;
	int	hour;
	int	minute;
	float	second;
	int	RETVAL;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 381 "UDUNITS.xs"
    {
	RETVAL = utCalendar(value, unit,
			    &year, &month, &day,
			    &hour, &minute, &second);
    }
#line 639 "UDUNITS.c"
	sv_setiv(ST(2), (IV)year);
	SvSETMAGIC(ST(2));
	sv_setiv(ST(3), (IV)month);
	SvSETMAGIC(ST(3));
	sv_setiv(ST(4), (IV)day);
	SvSETMAGIC(ST(4));
	sv_setiv(ST(5), (IV)hour);
	SvSETMAGIC(ST(5));
	sv_setiv(ST(6), (IV)minute);
	SvSETMAGIC(ST(6));
	sv_setnv(ST(7), (double)second);
	SvSETMAGIC(ST(7));
	ST(0) = sv_newmortal();
	sv_setiv(ST(0), (IV)RETVAL);
    }
    XSRETURN(1);
}

XS(XS_utUnitPtr_caltoval)
{
    dXSARGS;
    if (items != 7)
	croak("Usage: utUnitPtr::caltoval(unit, year, month, day, hour, minute, second)");
    {
	utUnit *	unit;
	int	year = (int)SvIV(ST(1));
	int	month = (int)SvIV(ST(2));
	int	day = (int)SvIV(ST(3));
	int	hour = (int)SvIV(ST(4));
	int	minute = (int)SvIV(ST(5));
	double	second = (double)SvNV(ST(6));
	double	RETVAL;

	if (sv_derived_from(ST(0), "utUnitPtr")) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not of type utUnitPtr");
#line 406 "UDUNITS.xs"
    {
	int	status = utInvCalendar(year, month, day,
				       hour, minute, second,
				       unit, &RETVAL);

	if (status == UT_EINVALID)
	    croak("not a unit of time");
	if (status == UT_ENOINIT)
	    croak("units module not initialized");
    }
#line 690 "UDUNITS.c"
	ST(0) = sv_newmortal();
	sv_setnv(ST(0), (double)RETVAL);
    }
    XSRETURN(1);
}

XS(XS_utUnitPtr_DESTROY)
{
    dXSARGS;
    if (items != 1)
	croak("Usage: utUnitPtr::DESTROY(unit)");
    {
	utUnit *	unit;

	if (SvROK(ST(0))) {
	    IV tmp = SvIV((SV*)SvRV(ST(0)));
	    unit = (utUnit *) tmp;
	}
	else
	    croak("unit is not a reference");
#line 424 "UDUNITS.xs"
    {
	free((char*)unit);
    }
#line 715 "UDUNITS.c"
    }
    XSRETURN_EMPTY;
}

#ifdef __cplusplus
extern "C"
#endif
XS(boot_UDUNITS)
{
    dXSARGS;
    char* file = __FILE__;

    XS_VERSION_BOOTCHECK ;

        newXS("UDUNITS::constant", XS_UDUNITS_constant, file);
        newXS("UDUNITS::init", XS_UDUNITS_init, file);
        newXS("UDUNITS::term", XS_UDUNITS_term, file);
        newXS("UDUNITS::new", XS_UDUNITS_new, file);
        newXS("UDUNITS::scan", XS_UDUNITS_scan, file);
        newXS("utUnitPtr::istime", XS_utUnitPtr_istime, file);
        newXS("utUnitPtr::hasorigin", XS_utUnitPtr_hasorigin, file);
        newXS("utUnitPtr::clear", XS_utUnitPtr_clear, file);
        newXS("utUnitPtr::dup", XS_utUnitPtr_dup, file);
        newXS("utUnitPtr::shift", XS_utUnitPtr_shift, file);
        newXS("utUnitPtr::scale", XS_utUnitPtr_scale, file);
        newXS("utUnitPtr::multiply", XS_utUnitPtr_multiply, file);
        newXS("utUnitPtr::invert", XS_utUnitPtr_invert, file);
        newXS("utUnitPtr::divide", XS_utUnitPtr_divide, file);
        newXS("utUnitPtr::raise", XS_utUnitPtr_raise, file);
        newXS("utUnitPtr::print", XS_utUnitPtr_print, file);
        newXS("utUnitPtr::convert", XS_utUnitPtr_convert, file);
        newXS("utUnitPtr::valtocal", XS_utUnitPtr_valtocal, file);
        newXS("utUnitPtr::caltoval", XS_utUnitPtr_caltoval, file);
        newXS("utUnitPtr::DESTROY", XS_utUnitPtr_DESTROY, file);
    XSRETURN_YES;
}

